Link against qdldl shared library

sebp opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->
- Disable downloading and compiling qdldl
- Only compile osqp shared library, not the static library
- Disable codegen which requires osqp compiled as static library

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/recipe.yaml```) and found it was in an excellent condition.

h-vetinari reviewed (commented):
Thanks for the PR! From a first look this LGTM, except the failures on windows obviously

sebp commented:
I managed to build the demo executables by exporting additional symbols. Unfortunately, building the tests binary still fails, presumably because it uses private API.

A workaround could be to extend the build process by building the static library to run the tests, and then re-build the shared library without tests, which finally is installed.

sebp committed:
Link against qdldl shared libary

- Only compile osqp shared library
- Disable codegen which requires osqp compiled as static library
- Add OSQP_USE_BUNDLED_QDLDL option: Runs tests with bundled QDLDL,
  but dynamically links against shared lib for package.

sebp committed:
MNT: Re-rendered with conda-smithy 3.52.3 and conda-forge-pinning 2025.10.10.13.43.27

sebp commented:
The build script now statically links against the bundled qdldl library for testing, and then re-builds the osqp lib and dynamically links it against qdldl for packaging.

When cross-compiling (macOS ARM package on Intel runner), the tests are skipped.

h-vetinari commented on recipe/recipe.yaml:
```diff
@@ -14,8 +14,8 @@ source:
     # - patches/0002-build-qdldl-interface-as-part-of-osqp.patch
       # install osqp_demo and osqp_tester executables to easily make them relocatable,
       # see: https://github.com/conda-forge/libosqp-feedstock/pull/12#discussion_r1807117589
```
the comment here is probably out of date now.

h-vetinari reviewed (commented):
Can you remind me why we cannot use the existing shared libraries with the demo for testing? Doing two different builds make the testing of the first one almost entirely pointless.

sebp commented:
> Can you remind me why we cannot use the existing shared libraries with the demo for testing? Doing two different builds make the testing of the first one almost entirely pointless.

The problem is the Windows build, because the tests use lots of symbols that aren't exported.

sebp committed:
Remove outdated comments

h-vetinari commented:
> The problem is the Windows build, because the tests use lots of symbols that aren't exported.

Can you please make a comment about this? Then I'm happy to merge.

Long-term we should consider marking the necessary symbols as `__declspec(dllexport)` so that this is not necessary anymore

sebp commented:
> Long-term we should consider marking the necessary symbols as `__declspec(dllexport)` so that this is not necessary anymore

I started to do this, but eventually gave up due to the sheer amount of symbols that require export.

sebp committed:
Add explanation for linking against bundled qdldl

h-vetinari reviewed (approved):
Thanks! :)
